File: seqspec/seqspec_info.py
import json
from argparse import ArgumentParser, Namespace, RawTextHelpFormatter
from pathlib import Path
from typing import Dict

from seqspec.Assay import Assay
from seqspec.utils import load_spec


def setup_info_args(parser) -> ArgumentParser:
    """Create and configure the info command subparser."""
    subparser = parser.add_parser(
        "info",
        description="""
Get information about spec.

Examples:
seqspec info -k modalities spec.yaml            # Get the list of modalities
seqspec info -f json spec.yaml                  # Get meta information in json format
seqspec info -f json -k library_spec spec.yaml  # Get library spec in json format
seqspec info -f json -k sequence_spec spec.yaml # Get sequence spec in json format
---
""",
        help="Get information from seqspec file",
        formatter_class=RawTextHelpFormatter,
    )

    subparser.add_argument("yaml", help="Sequencing specification yaml file", type=Path)
    choices = ["modalities", "meta", "sequence_spec", "library_spec"]
    subparser.add_argument(
        "-k",
        "--key",
        metavar="KEY",
        help=f"Object to display, [{', '.join(choices)}] (default: meta)",
        type=str,
        default="meta",
        required=False,
    )
    choices = ["tab", "json"]
    subparser.add_argument(
        "-f",
        "--format",
        metavar="FORMAT",
        help=f"The output format, [{', '.join(choices)}] (default: tab)",
        type=str,
        default="tab",
        required=False,
        choices=choices,
    )
    subparser.add_argument(
        "-o",
        "--output",
        metavar="OUT",
        help="Path to output file",
        type=Path,
        default=None,
        required=False,
    )
    return subparser


def validate_info_args(parser: ArgumentParser, args: Namespace) -> None:
    """Validate the info command arguments."""
    if not Path(args.yaml).exists():
        parser.error(f"Input file does not exist: {args.yaml}")

    if args.output and Path(args.output).exists() and not Path(args.output).is_file():
        parser.error(f"Output path exists but is not a file: {args.output}")


def run_info(parser: ArgumentParser, args: Namespace) -> None:
    """Run the info command."""
    validate_info_args(parser, args)

    spec = load_spec(args.yaml)

    if args.key:
        # Extract data
        info = seqspec_info(spec, args.key)
        # Format info
        result = format_info(info, args.key, args.format)

        if args.output:
            with open(args.output, "w") as f:
                if args.format == "json":
                    f.write(result)
                else:
                    print(result, file=f)
        else:
            print(result)


def seqspec_info(spec: Assay, key: str) -> Dict:
    """Get information from the spec based on the key.

    Args:
        spec: The Assay object to get info from
        key: The type of information to retrieve (modalities, meta, sequence_spec, library_spec)

    Returns:
        Dictionary containing the requested information

    Raises:
        KeyError: If the requested key is not supported
    """
    INFO_FUNCS = {
        "modalities": seqspec_info_modalities,
        "meta": seqspec_info_meta,
        "sequence_spec": seqspec_info_sequence_spec,
        "library_spec": seqspec_info_library_spec,
    }
    if key not in INFO_FUNCS:
        raise KeyError(
            f"Unsupported info key: {key}. Must be one of {list(INFO_FUNCS.keys())}"
        )
    return INFO_FUNCS[key](spec)


def format_info(info: Dict, key: str, fmt: str = "tab") -> str:
    """Format information based on the key and format.

    Args:
        info: Dictionary containing the information to format
        key: The type of information to format (modalities, meta, sequence_spec, library_spec)
        fmt: Output format (tab or json)

    Returns:
        Formatted string

    Raises:
        KeyError: If the requested key is not supported
    """
    FORMAT_FUNCS = {
        "modalities": format_modalities,
        "meta": format_meta,
        "sequence_spec": format_sequence_spec,
        "library_spec": format_library_spec,
    }
    if key not in FORMAT_FUNCS:
        raise KeyError(
            f"Unsupported format key: {key}. Must be one of {list(FORMAT_FUNCS.keys())}"
        )
    return FORMAT_FUNCS[key](info, fmt)


def seqspec_info_meta(spec: Assay) -> Dict:
    """Get meta information about the spec.

    Args:
        spec: The Assay object to get info from

    Returns:
        Dictionary containing meta information
    """
    sd = spec.to_dict()
    del sd["library_spec"]
    del sd["sequence_spec"]
    del sd["modalities"]
    return {"meta": sd}


def seqspec_info_library_spec(spec: Assay) -> Dict:
    """Get library specification information.

    Args:
        spec: The Assay object to get info from

    Returns:
        Dictionary containing library specifications by modality
    """
    modalities = spec.list_modalities()
    result = {}
    for m in modalities:
        libspec = spec.get_libspec(m)
        leaves = libspec.get_leaves()
        result[m] = leaves if leaves else []
    return {"library_spec": result}


def seqspec_info_sequence_spec(spec: Assay) -> Dict:
    """Get sequence specification information.

    Args:
        spec: The Assay object to get info from

    Returns:
        Dictionary containing sequence specifications
    """
    return {"sequence_spec": [i.model_dump() for i in spec.sequence_spec]}


def seqspec_info_modalities(spec: Assay) -> Dict:
    """Get list of modalities.

    Args:
        spec: The Assay object to get info from

    Returns:
        Dictionary containing list of modalities
    """
    return {"modalities": spec.list_modalities()}


def format_meta(info: Dict, fmt: str = "tab") -> str:
    """Format meta information.

    Args:
        info: Dictionary containing meta information from seqspec_info_meta
        fmt: Output format (tab or json)

    Returns:
        Formatted string
    """
    if fmt == "tab":
        return "\t".join(str(v) for v in info["meta"].values())
    elif fmt == "json":
        return json.dumps(info["meta"], sort_keys=False, indent=4)
    return ""


def format_modalities(info: Dict, fmt: str = "tab") -> str:
    """Format list of modalities.

    Args:
        info: Dictionary containing modalities from seqspec_info_modalities
        fmt: Output format (tab or json)

    Returns:
        Formatted string
    """
    if fmt == "tab":
        return "\t".join(info["modalities"])
    elif fmt == "json":
        return json.dumps(info["modalities"], sort_keys=False, indent=4)
    return ""


def format_sequence_spec(info: Dict, fmt: str = "tab") -> str:
    """Format sequence specification.

    Args:
        info: Dictionary containing sequence specs from seqspec_info_sequence_spec
        fmt: Output format (tab or json)

    Returns:
        Formatted string
    """
    if fmt == "tab":
        lines = []
        for r in info["sequence_spec"]:
            files = ",".join([i["file_id"] for i in r["files"]]) if r["files"] else ""
            lines.append(
                f"{r['modality']}\t{r['read_id']}\t{r['strand']}\t{r['min_len']}\t{r['max_len']}\t{r['primer_id']}\t{r['name']}\t{files}"
            )
        return "\n".join(lines)
    elif fmt == "json":
        return json.dumps(
            [i.model_dump() for i in info["sequence_spec"]], sort_keys=False, indent=4
        )
    return ""


def format_library_spec(info: Dict, fmt: str = "tab") -> str:
    """Format library specification.

    Args:
        info: Dictionary containing library specs from seqspec_info_library_spec
        fmt: Output format (tab or json)

    Returns:
        Formatted string
    """
    if fmt == "tab":
        lines = []
        for modality, regions in info["library_spec"].items():
            for r in regions:
                file = r["onlist"]["filename"] if r["onlist"] else None
                lines.append(
                    f"{modality}\t{r['region_id']}\t{r['region_type']}\t{r['name']}\t{r['sequence_type']}\t{r['sequence']}\t{r['min_len']}\t{r['max_len']}\t{file}"
                )
        return "\n".join(lines)
    elif fmt == "json":
        return json.dumps(
            {m: [i.model_dump() for i in r] for m, r in info["library_spec"].items()},
            sort_keys=False,
            indent=4,
        )
    return ""

File: src/seqspec_info.rs
use crate::auth::RemoteAccess;
use crate::models::assay::Assay;
use crate::models::read::Read;
use crate::models::region::Region;
use crate::utils;
use clap::Args;
use serde_json::{json, Value};
use std::collections::BTreeMap;
use std::fs;
use std::io::Write;
use std::path::PathBuf;

#[derive(Debug, Args)]
pub struct InfoArgs {
    #[clap(help = "Path or URL to sequencing specification YAML", required = true)]
    yaml: String,

    #[clap(
        short,
        long,
        help = "Object to display",
        value_name = "KEY",
        default_value = "meta",
        value_parser = ["modalities", "meta", "sequence_spec", "library_spec"]
    )]
    key: String,

    #[clap(
        short,
        long,
        help = "The output format",
        value_name = "FORMAT",
        default_value = "tab",
        value_parser = ["tab", "json"]
    )]
    format: String,

    #[clap(short, long, help = "Path to output file", value_name = "OUT")]
    output: Option<PathBuf>,

    #[clap(long, env = "SEQSPEC_AUTH_PROFILE", value_name = "PROFILE")]
    auth_profile: Option<String>,
}

pub fn run_info(args: &InfoArgs) {
    let remote_access = RemoteAccess::load(args.auth_profile.as_deref()).unwrap_or_else(|err| {
        eprintln!("{}", err);
        std::process::exit(1);
    });
    validate_info_args(args, &remote_access);
    let spec = utils::load_spec_source(&args.yaml, &remote_access).unwrap_or_else(|err| {
        eprintln!("{}", err);
        std::process::exit(1);
    });

    let info = seqspec_info(&spec, &args.key);
    let result = format_info(&spec, info, &args.key, &args.format);

    if let Some(out) = &args.output {
        let mut f = fs::File::create(out).unwrap();
        writeln!(f, "{}", result).unwrap();
    } else {
        println!("{}", result);
    }
}

fn validate_info_args(args: &InfoArgs, remote_access: &RemoteAccess) {
    if let Err(err) = utils::validate_source_exists(&args.yaml, remote_access) {
        eprintln!("{}", err);
        std::process::exit(1);
    }
    if let Some(out) = &args.output {
        if out.exists() && !out.is_file() {
            eprintln!("Output path exists but is not a file: {}", out.display());
            std::process::exit(1);
        }
    }
}

// ---------------- Core info ----------------

#[derive(Clone)]
enum InfoData {
    Modalities(Vec<String>),
    Meta(Value),
    SequenceSpec(Vec<Read>),
    LibrarySpec(BTreeMap<String, Vec<Region>>),
}

fn seqspec_info(spec: &Assay, key: &str) -> InfoData {
    match key {
        "modalities" => InfoData::Modalities(spec.list_modalities()),
        "meta" => InfoData::Meta(seqspec_info_meta(spec)),
        "sequence_spec" => InfoData::SequenceSpec(spec.sequence_spec.clone()),
        "library_spec" => InfoData::LibrarySpec(seqspec_info_library_spec(spec)),
        _ => panic!("Unsupported info key: {}", key),
    }
}

fn seqspec_info_meta(spec: &Assay) -> Value {
    // preserve a logical, stable order of fields similar to Python's model_dump
    let mut m = serde_json::Map::new();
    if let Some(v) = &spec.seqspec_version {
        m.insert("seqspec_version".to_string(), json!(v));
    }
    m.insert("assay_id".to_string(), json!(spec.assay_id));
    m.insert("name".to_string(), json!(spec.name));
    m.insert("doi".to_string(), json!(spec.doi));
    m.insert("date".to_string(), json!(spec.date));
    m.insert("description".to_string(), json!(spec.description));
    m.insert("lib_struct".to_string(), json!(spec.lib_struct));
    if let Some(v) = &spec.library_kit {
        m.insert("library_kit".to_string(), json!(v));
    }
    if let Some(v) = &spec.library_protocol {
        m.insert("library_protocol".to_string(), json!(v));
    }
    if let Some(v) = &spec.sequence_kit {
        m.insert("sequence_kit".to_string(), json!(v));
    }
    if let Some(v) = &spec.sequence_protocol {
        m.insert("sequence_protocol".to_string(), json!(v));
    }
    Value::Object(m)
}

fn seqspec_info_library_spec(spec: &Assay) -> BTreeMap<String, Vec<Region>> {
    let mut result: BTreeMap<String, Vec<Region>> = BTreeMap::new();
    for m in spec.list_modalities() {
        if let Some(libspec) = spec.get_libspec(&m) {
            let leaves = libspec.get_leaves();
            result.insert(m, leaves);
        }
    }
    result
}

// ---------------- Formatting ----------------

fn format_info(_spec: &Assay, info: InfoData, key: &str, fmt: &str) -> String {
    match (key, fmt) {
        ("modalities", "tab") => format_modalities_tab(&info),
        ("modalities", "json") => format_modalities_json(&info),
        ("meta", "tab") => format_meta_tab(&info),
        ("meta", "json") => format_meta_json(&info),
        ("sequence_spec", "tab") => format_sequence_spec_tab(&info),
        ("sequence_spec", "json") => format_sequence_spec_json(&info),
        ("library_spec", "tab") => format_library_spec_tab(&info),
        ("library_spec", "json") => format_library_spec_json(&info),
        _ => String::new(),
    }
}

fn format_modalities_tab(info: &InfoData) -> String {
    if let InfoData::Modalities(v) = info {
        v.join("\t")
    } else {
        String::new()
    }
}
fn format_modalities_json(info: &InfoData) -> String {
    if let InfoData::Modalities(v) = info {
        serde_json::to_string_pretty(v).unwrap()
    } else {
        String::new()
    }
}

fn format_meta_tab(info: &InfoData) -> String {
    if let InfoData::Meta(v) = info {
        let obj = v.as_object().unwrap();
        let mut vals: Vec<String> = Vec::new();
        for k in [
            "seqspec_version",
            "assay_id",
            "name",
            "doi",
            "date",
            "description",
            "lib_struct",
            "library_kit",
            "library_protocol",
            "sequence_kit",
            "sequence_protocol",
        ] {
            if let Some(val) = obj.get(k) {
                vals.push(if val.is_null() {
                    String::new()
                } else {
                    val.to_string().trim_matches('"').to_string()
                });
            }
        }
        vals.join("\t")
    } else {
        String::new()
    }
}
fn format_meta_json(info: &InfoData) -> String {
    if let InfoData::Meta(v) = info {
        serde_json::to_string_pretty(v).unwrap()
    } else {
        String::new()
    }
}

fn format_sequence_spec_tab(info: &InfoData) -> String {
    if let InfoData::SequenceSpec(reads) = info {
        let mut lines: Vec<String> = Vec::new();
        for r in reads {
            let files = if r.files.is_empty() {
                String::new()
            } else {
                r.files
                    .iter()
                    .map(|f| f.file_id.clone())
                    .collect::<Vec<_>>()
                    .join(",")
            };
            lines.push(format!(
                "{}\t{}\t{}\t{}\t{}\t{}\t{}\t{}",
                r.modality, r.read_id, r.strand, r.min_len, r.max_len, r.primer_id, r.name, files
            ));
        }
        lines.join("\n")
    } else {
        String::new()
    }
}
fn format_sequence_spec_json(info: &InfoData) -> String {
    if let InfoData::SequenceSpec(reads) = info {
        serde_json::to_string_pretty(reads).unwrap()
    } else {
        String::new()
    }
}

fn format_library_spec_tab(info: &InfoData) -> String {
    if let InfoData::LibrarySpec(map) = info {
        let mut lines: Vec<String> = Vec::new();
        for (modality, regions) in map {
            for r in regions {
                let file = r
                    .onlist
                    .as_ref()
                    .map(|o| o.filename.clone())
                    .unwrap_or_else(|| "None".to_string());
                lines.push(format!(
                    "{}\t{}\t{}\t{}\t{}\t{}\t{}\t{}\t{}",
                    modality,
                    r.region_id,
                    r.region_type,
                    r.name,
                    r.sequence_type,
                    r.sequence,
                    r.min_len,
                    r.max_len,
                    file
                ));
            }
        }
        lines.join("\n")
    } else {
        String::new()
    }
}
fn format_library_spec_json(info: &InfoData) -> String {
    if let InfoData::LibrarySpec(map) = info {
        serde_json::to_string_pretty(map).unwrap()
    } else {
        String::new()
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::utils::load_spec;

    fn dogma_spec() -> Assay {
        load_spec(&PathBuf::from("tests/fixtures/spec.yaml"))
    }

    #[test]
    fn test_info_modalities() {
        let spec = dogma_spec();
        let info = seqspec_info(&spec, "modalities");
        if let InfoData::Modalities(v) = info {
            assert!(v.contains(&"rna".to_string()));
            assert!(v.contains(&"atac".to_string()));
            assert_eq!(v.len(), 4);
        } else {
            panic!("Expected Modalities variant");
        }
    }

    #[test]
    fn test_info_meta() {
        let spec = dogma_spec();
        let info = seqspec_info(&spec, "meta");
        if let InfoData::Meta(v) = info {
            let obj = v.as_object().unwrap();
            assert!(obj.contains_key("assay_id"));
            assert!(obj.contains_key("name"));
            assert_eq!(obj["assay_id"].as_str().unwrap(), "DOGMAseq-DIG");
        } else {
            panic!("Expected Meta variant");
        }
    }

    #[test]
    fn test_info_sequence_spec() {
        let spec = dogma_spec();
        let info = seqspec_info(&spec, "sequence_spec");
        if let InfoData::SequenceSpec(reads) = info {
            assert_eq!(reads.len(), 9); // 2 RNA + 3 ATAC + 2 Protein + 2 Tag
            let rna_reads: Vec<_> = reads.iter().filter(|r| r.modality == "rna").collect();
            assert_eq!(rna_reads.len(), 2);
            let atac_reads: Vec<_> = reads.iter().filter(|r| r.modality == "atac").collect();
            assert_eq!(atac_reads.len(), 3);
        } else {
            panic!("Expected SequenceSpec variant");
        }
    }

    #[test]
    fn test_info_library_spec() {
        let spec = dogma_spec();
        let info = seqspec_info(&spec, "library_spec");
        if let InfoData::LibrarySpec(map) = info {
            assert_eq!(map.len(), 4);
            assert_eq!(map["rna"].len(), 5); // 5 RNA leaf regions
            assert_eq!(map["rna"][0].region_id, "rna_truseq_read1");
        } else {
            panic!("Expected LibrarySpec variant");
        }
    }

    #[test]
    fn test_format_modalities_tab() {
        let spec = dogma_spec();
        let info = seqspec_info(&spec, "modalities");
        let result = format_info(&spec, info, "modalities", "tab");
        let parts: Vec<&str> = result.split('\t').collect();
        assert_eq!(parts.len(), 4);
        assert!(parts.contains(&"rna"));
        assert!(parts.contains(&"atac"));
        assert!(parts.contains(&"protein"));
        assert!(parts.contains(&"tag"));
    }

    #[test]
    fn test_format_meta_json() {
        let spec = dogma_spec();
        let info = seqspec_info(&spec, "meta");
        let result = format_info(&spec, info, "meta", "json");
        let parsed: Value = serde_json::from_str(&result).unwrap();
        assert!(parsed.is_object());
        assert_eq!(parsed["assay_id"].as_str().unwrap(), "DOGMAseq-DIG");
    }

    #[test]
    fn test_format_meta_tab() {
        let spec = dogma_spec();
        let info = seqspec_info(&spec, "meta");
        let result = format_info(&spec, info, "meta", "tab");
        let parts: Vec<&str> = result.split('\t').collect();
        // Tab format includes: version, assay_id, name, doi, date, description, lib_struct, ...
        assert!(parts.len() >= 7);
        assert!(parts.contains(&"DOGMAseq-DIG")); // assay_id
        assert!(parts.contains(&"DOGMAseq-DIG")); // name too
    }

    #[test]
    fn test_format_sequence_spec_tab() {
        let spec = dogma_spec();
        let info = seqspec_info(&spec, "sequence_spec");
        let result = format_info(&spec, info, "sequence_spec", "tab");
        let lines: Vec<&str> = result.lines().collect();
        assert_eq!(lines.len(), 9); // 9 reads total
                                    // First line should be an RNA read
        assert!(
            lines[0].starts_with("rna\t")
                || lines[0].starts_with("protein\t")
                || lines[0].starts_with("tag\t")
                || lines[0].starts_with("atac\t")
        );
        // Check that rna and atac both appear
        let rna_lines = lines.iter().filter(|l| l.starts_with("rna\t")).count();
        assert_eq!(rna_lines, 2);
        let atac_lines = lines.iter().filter(|l| l.starts_with("atac\t")).count();
        assert_eq!(atac_lines, 3);
    }

    #[test]
    fn test_format_sequence_spec_json() {
        let spec = dogma_spec();
        let info = seqspec_info(&spec, "sequence_spec");
        let result = format_info(&spec, info, "sequence_spec", "json");
        let parsed: Value = serde_json::from_str(&result).unwrap();
        let arr = parsed.as_array().unwrap();
        assert_eq!(arr.len(), 9);
    }

    #[test]
    fn test_format_library_spec_tab() {
        let spec = dogma_spec();
        let info = seqspec_info(&spec, "library_spec");
        let result = format_info(&spec, info, "library_spec", "tab");
        let lines: Vec<&str> = result.lines().collect();
        // Count lines per modality
        let rna_lines = lines.iter().filter(|l| l.starts_with("rna\t")).count();
        assert_eq!(rna_lines, 5); // 5 RNA leaf regions
    }

    #[test]
    fn test_format_library_spec_json() {
        let spec = dogma_spec();
        let info = seqspec_info(&spec, "library_spec");
        let result = format_info(&spec, info, "library_spec", "json");
        let parsed: Value = serde_json::from_str(&result).unwrap();
        let obj = parsed.as_object().unwrap();
        assert_eq!(obj.len(), 4);
        assert_eq!(obj["rna"].as_array().unwrap().len(), 5);
    }

    #[test]
    fn test_format_modalities_json() {
        let spec = dogma_spec();
        let info = seqspec_info(&spec, "modalities");
        let result = format_info(&spec, info, "modalities", "json");
        let parsed: Value = serde_json::from_str(&result).unwrap();
        let arr = parsed.as_array().unwrap();
        assert_eq!(arr.len(), 4);
    }
}
